### PR TITLE
Support sorted-merge GroupPartitionsExec on GPU [databricks][fast-ut][reduced-it]

### LINKED ISSUE #15602: [FEA] Support GroupPartitionsExec on GPU
**Is your feature request related to a problem? Please describe.**

Apache Spark 4.2 introduced `GroupPartitionsExec` as part of the Storage Partition Join (SPJ) refactor. It aligns partitions with the same partition key without a shuffle. The RAPIDS Accelerator currently needs to preserve the original CPU subtree when this operator falls back; otherwise GPU child conversion can lose the required `KeyedPartitioning`. Although this fallback is correct, it can move a large SPJ subtree to CPU and cause a significant performance penalty.

This issue provides dedicated tracking for GPU support. The Spark 4.2 API audit and the initial fallback work are tracked by #15307 and #15555.

**Describe the solution you'd like**

Implement support in two phases, separated by whether Spark enables sorted merge.

### Phase 1: GPU support without sorted merge

- [x] Add a lightweight `GpuGroupPartitionsExec` for `enableSortedMerge = false`.
- [x] Preserve all Spark grouping parameters: join-key projection, expected partition keys, reducers, split distribution, and empty/padded groups.
- [x] Reuse Spark's grouping semantics and `GroupedPartitionCoalescer` rather than duplicating `KeyedPartitioning` projection/reduction/alignment logic.
- [x] Execute columnar children through `CoalescedRDD`, concatenating the selected parent partitions sequentially without a shuffle.
- [x] Allow batch coalescing after the operator because one output partition can contain batches from multiple input partitions.
- [x] Keep the existing full CPU-subtree fallback for sorted-merge plans.
- [x] Add tests for basic grouping, multiple splits per key, missing keys/empty padding, shuffle-free SPJ, and sorted-merge fallback.

#### How Spark's CPU implementation works

For normal grouping, Spark uses `CoalescedRDD` to read grouped parent partitions sequentially. Sequential concatenation does not preserve full ordering. For example, concatenating sorted runs `[1, 3]` and `[2, 5]` produces `[1, 3, 2, 5]`.

When all of the following conditions hold, `EnsureRequirements` can set `enableSortedMerge = true` on `GroupPartitionsExec`:

- `spark.sql.sources.v2.bucketing.preserveOrderingOnCoalesce.enabled` is enabled.
- Multiple partitions are actually coalesced.
- The child reports a non-empty `outputOrdering`.
- The child subtree implements `SafeForKWayMerge`.
- A parent operator requires the ordering and would otherwise need a `SortExec`.

The CPU row path then uses `SortedMergeCoalescedRDD[InternalRow]`. It opens the sorted parent iterators and performs a priority-queue-based k-way merge, preserving the complete child ordering. In this mode `GroupPartitionsExec.supportsColumnar` is false. This can eliminate a redundant sort before operators such as `SortMergeJoinExec`.

#### GPU work

- [ ] Convert the complete Spark `SortOrder`, including direction and null ordering, to GPU ordering expressions.
- [ ] Implement a spillable, retry-safe columnar merge that emits target-sized batches instead of one large table.
- [ ] Preserve `outputOrdering` only when the GPU execution path actually guarantees it.
- [ ] Manage multiple parent iterators and all GPU resources safely across task completion, exceptions, and OOM retry.
- [ ] Define supported sort-key/data types and fall back safely where cuDF merge has limitations.
- [ ] Add correctness tests for interleaved sorted runs, ASC/DESC, null ordering, secondary keys, empty/single partitions, spilling, and resource cleanup.

A correctness-first implementation could reuse `GpuSortExec` after ordinary partition coalescing. A more efficient implementation would reuse cuDF `Table.merge` and the existing out-of-core sort infrastructure to merge already-sorted batches without sorting all rows again.

**Describe alternatives you've considered**

- Keep the entire operator subtree on CPU. This is correct but can have a large performance cost.
- When a CPU `SortMergeJoinExec` will be replaced by a GPU hash join, clear the now-unnecessary sorted-merge requirement and use Phase 1. This should cover many RAPIDS join plans but does not support genuinely order-dependent parents.
- Always concatenate and then run a full GPU sort. This is simpler and safe but does not exploit already-sorted input runs.

**Additional context**

- Existing tracking: #15307
- Correct CPU fallback: #15555
- Spark SPJ refactor: https://github.com/apache/spark/pull/54330
- Spark ordering preservation: https://issues.apache.org/jira/browse/SPARK-55715
- Dynamic sorted-merge enablement: https://issues.apache.org/jira/browse/SPARK-56549

### PR DESCRIPTION
**Pull request by firestarman**
Fixes #15602.

### Description
- Converts Spark 4.2 sorted-merge `GroupPartitionsExec` plans to GPU so storage-partitioned joins no longer fall back to the CPU when partition coalescing must preserve ordering.
- Coalesces each partition group without a shuffle and uses the existing spillable `GpuOutOfCoreSortIterator` because it already provides GPU OOM retry, spill handling, ordering semantics, and target-sized output batches.
- Sorts only groups containing multiple independent input partitions because empty and single-input groups already preserve their reported ordering.
- Preserves the original CPU subtree when normal GPU compatibility checks reject replacement because CPU `GroupPartitionsExec` reads its child's keyed partitioning at execution time.
- Adds Spark 4.2 unit and integration coverage for ascending and descending ordering, non-default null ordering, padded and single-input groups, target-sized batches, GPU OOM retry, fallback safety, metrics, batching, and child output type propagation.
- Validated with `mvn -s ~/.m2/settings_art.xml -f scala2.13/pom.xml -Dbuildver=420 -Dcuda.version=cuda13 -DwildcardSuites=com.nvidia.spark.rapids.GroupPartitionsExecSuite -pl tests test` (15 tests passed), Spark 4.2 `test_group_partitions_sorted_merge`, Spark 4.1.3 `SortExecSuite` (4 tests passed), and the cross-version partial-clustering DISTINCT integration test.
- Performance compared latest `NVDA/main` commit `85604957c` using CPU sorted merge with this PR commit `e5fc8d90b` using GPU full sort. Both variants used Spark 4.2, Scala 2.13, `local[4]`, an 8 GiB driver heap, one concurrent GPU task, AQE and broadcast joins disabled, four keys, and 16 million rows per join side.
- Performance runs used fresh Spark processes in the order main/PR/main/PR. Each process ran two warm-ups followed by five measured actions, producing ten measured actions per variant and scenario.
- Performance query:

```sql
SELECT /*+ MERGE(l, r) */ count(*) AS matched_rows
FROM group_partitions_perf_left l
JOIN group_partitions_perf_right r
  ON l.id = r.id AND l.value = r.value
```

| Scenario | `NVDA/main` CPU median | PR GPU median | Relative performance |
| --- | ---: | ---: | ---: |
| 8 streams/key, 500,000 rows/stream | 7.752 s | 4.940 s | 1.569x |
| 32 streams/key, 125,000 rows/stream | 32.641 s | 35.237 s | 0.926x |

- The performance results show that GPU full sort provides a clear benefit at moderate fan-in, while the 32-stream result identifies high fan-in as its weak point.
- CPU sorted merge combines already sorted streams with work closer to `O(N log K)`, whereas this correctness-first GPU implementation re-sorts the full group with work closer to `O(N log N)`. The high-fan-in result motivates evaluating true GPU k-way merge in that scenario; fan-in alone is not a sufficient fallback heuristic because the crossover also depends on row count, row width, and ordering keys.

### Checklists

Documentation
- [ ] Updated for new or modified user-facing features or behaviors
- [x] No user-facing change

Testing
- [x] Added or modified tests to cover new code paths
- [ ] Covered by existing tests
      (Please provide the names of the existing tests in the PR description.)
- [ ] Not required

Performance
- [x] Tests ran and results are added in the PR description
- [ ] Issue filed with a link in the PR description
- [ ] Not required

### CODE REVIEW

**Comment by firestarman** [comment]
Finished the manual check, now ready for review.

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
build

File: integration_tests/src/main/python/group_partitions_test.py [hunk 1]
```diff
@@ -124,3 +147,47 @@ def distinct_after_spj(spark):
...
+
+
+@allow_non_gpu("BatchScanExec")
+@ignore_order(local=True)
+@pytest.mark.skipif(not is_spark_420_or_later(), reason="GroupPartitionsExec requires Spark 4.2")
+def test_group_partitions_sorted_merge():
+    def sorted_merge_spj(spark):
+        source = \
+            "com.nvidia.spark.rapids.tests.datasourcev2.GroupPartitionsSortedDataSource"
+        spark.read.format(source) \
```
Comment (firestarman) [inline integration_tests/src/main/python/group_partitions_test.py]: Could the identical left/right source-registration chains be expressed as a loop over `("left", "right")`? Keeping the shared direction and null-ordering options in one place prevents the two inputs from drifting during later edits.

File: sql-plugin/src/main/spark420/scala/com/nvidia/spark/rapids/shims/GpuGroupPartitionsExec.scala [hunk 2]
```diff
@@ -144,11 +189,49 @@ case class GpuGroupPartitionsExec(
...
       sparkContext.emptyRDD
     } else {
+      // Preserve Spark's exact parent-to-output partition mapping without introducing a shuffle.
       val partitionCoalescer = new GroupedPartitionCoalescer(partitionGroups)
-      child.executeColumnar().coalesce(
+      val coalesced = child.executeColumnar().coalesce(
         partitionGroups.size,
         shuffle = false,
         Some(partitionCoalescer))
+      if (needsSortedMerge) {
```
Comment (firestarman) [inline sql-plugin/src/main/spark420/scala/com/nvidia/spark/rapids/shims/GpuGroupPartitionsExec.scala:238]: Could the sorted-merge branch be extracted into a private helper that accepts the coalesced RDD? Keeping sorter setup, per-partition ownership handling, metrics, and the task-completion callback together would make the execution branches easier to follow.

File: tests/src/test/spark420/scala/com/nvidia/spark/rapids/GroupPartitionsExecSuite.scala [hunk 3]
```diff
@@ -181,19 +198,262 @@ class GroupPartitionsExecSuite extends SparkQueryCompareTestSuite {
...
+          .map(_.getLong(0))
+          .toSeq
+        assert(actual == expected)
+        assert(groupPartitions.metrics(GpuMetric.SORT_TIME).value > 0)
+        assert(groupPartitions.allMetrics(GpuMetric.NUM_OUTPUT_ROWS).value == expected.size)
+      }
+    }
+  }
+
+  test("GpuGroupPartitionsExec sorted merge handles padded and single-input groups") {
```
Comment (firestarman) [inline tests/src/test/spark420/scala/com/nvidia/spark/rapids/GroupPartitionsExecSuite.scala:345]: Could this test run output partition 2 in isolation with a fresh plan and assert that `SORT_TIME` remains zero? The current aggregate assertion stays positive because the multi-input partition sorts, so it would not catch the singleton fast path being routed through the sorter.

File: tests/src/test/spark420/scala/com/nvidia/spark/rapids/GroupPartitionsExecSuite.scala [hunk 4]
```diff
@@ -181,19 +198,262 @@ class GroupPartitionsExecSuite extends SparkQueryCompareTestSuite {
...
+        .map(_._2)
+        .toSeq
+      assert(actual == Seq(Seq.empty, Seq(1, 2, 3, 4), Seq(5, 6)))
+      assert(groupPartitions.allMetrics(GpuMetric.NUM_OUTPUT_ROWS).value == 6)
+      assert(groupPartitions.allMetrics(GpuMetric.NUM_OUTPUT_BATCHES).value == 2)
+      assert(groupPartitions.allMetrics(GpuMetric.SORT_TIME).value > 0)
+    }
+  }
+
+  test("GpuGroupPartitionsExec sorted merge emits target-sized batches") {
```
Comment (firestarman) [inline tests/src/test/spark420/scala/com/nvidia/spark/rapids/GroupPartitionsExecSuite.scala:397]: Could this test collect and assert the device byte size of each emitted batch against the configured 16-KiB target, closing batches in `finally`? `NUM_OUTPUT_BATCHES > 1` still passes if the target is ignored or inflated enough to emit several oversized batches.

File: tests/src/test/spark420/scala/com/nvidia/spark/rapids/GroupPartitionsExecSuite.scala [hunk 5]
```diff
@@ -237,3 +497,33 @@ class GroupPartitionsExecSuite extends SparkQueryCompareTestSuite {
...
   }
 }
+
+private case class GroupPartitionsTestGpuLeaf(
+    override val output: Seq[Attribute],
+    partitionValues: Seq[Seq[Integer]],
+    override val outputOrdering: Seq[SortOrder],
+    injectRetry: Boolean = false)
+    extends ShimLeafExecNode with GpuExec {
+  override def outputPartitioning = UnknownPartitioning(partitionValues.size)
```
Comment (firestarman) [inline tests/src/test/spark420/scala/com/nvidia/spark/rapids/GroupPartitionsExecSuite.scala]: Could `outputPartitioning` declare an explicit `Partitioning` return type? This keeps the overridden public plan contract explicit instead of exposing the concrete inferred `UnknownPartitioning` type.

File: integration_tests/src/main/python/group_partitions_test.py (hunk 1 above)
Comment (firestarman) [reply integration_tests/src/main/python/group_partitions_test.py]: Addressed in e99ddf174. The left and right sources are now registered through one loop so shared options stay in sync.

File: tests/src/test/spark420/scala/com/nvidia/spark/rapids/GroupPartitionsExecSuite.scala (hunk 4 above)
Comment (firestarman) [reply tests/src/test/spark420/scala/com/nvidia/spark/rapids/GroupPartitionsExecSuite.scala:397]: Addressed in e99ddf174. The test now records each emitted batch device size, closes every batch in finally, and checks all batches against the 16-KiB target.

File: tests/src/test/spark420/scala/com/nvidia/spark/rapids/GroupPartitionsExecSuite.scala (hunk 3 above)
Comment (firestarman) [reply tests/src/test/spark420/scala/com/nvidia/spark/rapids/GroupPartitionsExecSuite.scala:345]: Addressed in e99ddf174. The test creates a fresh plan, runs output partition 2 alone, verifies its rows, and asserts SORT_TIME remains zero.

File: tests/src/test/spark420/scala/com/nvidia/spark/rapids/GroupPartitionsExecSuite.scala (hunk 5 above)
Comment (firestarman) [reply tests/src/test/spark420/scala/com/nvidia/spark/rapids/GroupPartitionsExecSuite.scala]: Addressed in e99ddf174. outputPartitioning now declares the explicit Partitioning return type.

File: sql-plugin/src/main/spark420/scala/com/nvidia/spark/rapids/shims/GpuGroupPartitionsExec.scala (hunk 2 above)
Comment (firestarman) [reply sql-plugin/src/main/spark420/scala/com/nvidia/spark/rapids/shims/GpuGroupPartitionsExec.scala:238]: Addressed in e99ddf174. The sorted-merge setup and per-partition ownership logic now live in sortCoalescedPartitions.

**Comment by firestarman** [comment]
build

File: integration_tests/src/main/spark420/scala/com/nvidia/spark/rapids/tests/datasourcev2/GroupPartitionsSortedDataSource.scala [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*** spark-rapids-shim-json-lines
+{"spark": "420"}
```
Comment (wjxiz1992) [inline integration_tests/src/main/spark420/scala/com/nvidia/spark/rapids/tests/datasourcev2/GroupPartitionsSortedDataSource.scala:18]: [P2] Include this test data source in the Spark 5.0 shim

Current main now enables the production `GpuGroupPartitionsExec` source for both Spark 4.2 and 5.0, but this new provider remains 4.2-only while `test_group_partitions_sorted_merge` is guarded by `is_spark_420_or_later()`. A clean merge-tree against current main therefore selects the Python test for Spark 5.0 without packaging this provider; provider loading is expected to fail. I did not run the Spark 5.0 IT.

Please add `{"spark": "500"}` here (and preferably to the focused Scala suite) and validate the Spark-master build, or narrow the Python guard if Spark 5.0 coverage is intentionally out of scope.

**Review by wjxiz1992 — COMMENTED** [review]
One Spark-master shim-coverage issue to address when updating this branch onto current main.

File: integration_tests/src/main/spark420/scala/com/nvidia/spark/rapids/tests/datasourcev2/GroupPartitionsSortedDataSource.scala [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+    }
+
+    side match {
+      case "left" =>
+        Array(
+          rows(1, null, 10, 30),
+          rows(1, null, 20, 40),
+          rows(2, 5, 25),
+          rows(3, 15))
+      case "right" =>
```
Comment (res-life) [inline integration_tests/src/main/spark420/scala/com/nvidia/spark/rapids/tests/datasourcev2/GroupPartitionsSortedDataSource.scala:67]: Could the right-side key-1 values use a different per-partition distribution from the left side? Both sides currently produce the same concatenated secondary-key sequence, so a regression that sorts only by id and drops the value ordering could still let the sort-merge join produce the expected result.

For example, keep each right-side partition individually sorted while changing the distribution:

```scala
case "right" =>
  Array(
    rows(1, null, 10, 20),
    rows(1, null, 30, 40),
    rows(2, 5, 25),
    rows(3, 15))
```

Both sides retain the same rows, but their unsorted concatenations differ, making the integration test sensitive to loss of the second `SortOrder`.

File: tests/src/test/spark420/scala/com/nvidia/spark/rapids/GroupPartitionsExecSuite.scala [hunk 8]
```diff
@@ -181,19 +199,284 @@ class GroupPartitionsExecSuite extends SparkQueryCompareTestSuite {
...
+    val childMeta = mock(classOf[SparkPlanMeta[SparkPlan]])
+    when(childMeta.availableRuntimeDataTransition).thenReturn(true)
+    when(childMeta.outputAttributes).thenReturn(Seq(convertedAttr))
+    val meta = spy(originalMeta)
+    doReturn(Seq(childMeta)).when(meta).childPlans
+
+    assert(meta.outputAttributes == Seq(convertedAttr))
+  }
+
+  test("GroupPartitionsExec fallback keeps the original CPU subtree") {
```
Comment (res-life) [inline tests/src/test/spark420/scala/com/nvidia/spark/rapids/GroupPartitionsExecSuite.scala]: Could this fallback test make the rejection originate from an unsupported sorted-merge ordering expression? It currently calls `meta.willNotWorkOnGpu` directly, so it would still pass if `sortedMergeOrdering` stopped being exposed through `childExprs`.

One way to exercise the new path is to create an `enableSortedMerge = true` plan, stub its `outputOrdering` with an expression that has no GPU rule (for example `SortOrder(CurrentDatabase(), Ascending)`), and then assert after `wrapAndTagPlan` that:

```scala
assert(meta.childExprs.nonEmpty)
assert(!meta.canThisBeReplaced)
assert(meta.convertToCpu().eq(groupPartitions))
```

This would verify both expression-driven fallback and preservation of the original CPU subtree.

File: integration_tests/src/main/spark420/scala/com/nvidia/spark/rapids/tests/datasourcev2/GroupPartitionsSortedDataSource.scala (hunk 7 above)
Comment (firestarman) [reply integration_tests/src/main/spark420/scala/com/nvidia/spark/rapids/tests/datasourcev2/GroupPartitionsSortedDataSource.scala:67]: Addressed in 60b6f7072. The right-side key-1 rows now use a different per-partition distribution while preserving the same row set and per-partition ordering. The focused Spark 4.2 integration test passes.

File: integration_tests/src/main/spark420/scala/com/nvidia/spark/rapids/tests/datasourcev2/GroupPartitionsSortedDataSource.scala (hunk 6 above)
Comment (firestarman) [reply integration_tests/src/main/spark420/scala/com/nvidia/spark/rapids/tests/datasourcev2/GroupPartitionsSortedDataSource.scala:18]: Addressed in 60b6f7072 after merging current main. The data source and focused Scala suite now include the Spark 500 shim, and the production shim retained its 500 entry. Spark 500 validate/scalastyle passes. I also attempted the full build; it currently stops earlier in unmodified main code because GpuColumnVector imports VariantTypeShims while that shim excludes build 500.

File: tests/src/test/spark420/scala/com/nvidia/spark/rapids/GroupPartitionsExecSuite.scala (hunk 8 above)
Comment (firestarman) [reply tests/src/test/spark420/scala/com/nvidia/spark/rapids/GroupPartitionsExecSuite.scala]: Addressed in 60b6f7072. The test now uses an unsupported CurrentDatabase sorted-merge ordering, asserts childExprs is populated, and verifies expression-driven rejection preserves the original CPU subtree. GroupPartitionsExecSuite passes all 15 tests.

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
build

File: sql-plugin/src/main/spark420/scala/com/nvidia/spark/rapids/shims/GpuGroupPartitionsExec.scala [hunk 9]
```diff
@@ -40,21 +41,34 @@ class GpuGroupPartitionsExecMeta(
...
+  override protected val useOutputAttributesOfChild: Boolean = true
+
+  // This plan is transparent to row-to-columnar transitions inserted above its child.
+  override val availableRuntimeDataTransition: Boolean =
+    childPlans.head.availableRuntimeDataTransition
+
+  private val sortedMergeOrdering: Seq[BaseExprMeta[SortOrder]] =
     if (groupPartitions.enableSortedMerge) {
-      willNotWorkOnGpu("Sorted-merge GroupPartitionsExec is not supported on GPU")
+      groupPartitions.outputOrdering.map(GpuOverrides.wrapExpr(_, conf, Some(this)))
```
Comment (res-life) [inline sql-plugin/src/main/spark420/scala/com/nvidia/spark/rapids/shims/GpuGroupPartitionsExec.scala]: [P1] Drop planner-only `sameOrderExpressions` before GPU conversion

Please strip `sameOrderExpressions` from each `SortOrder` before wrapping it for GPU execution. Apache Spark made the equivalent change in SPARK-58324, commit [`fb34cbeeec8a041a6e8a96cc60e962e78116a19f`](https://github.com/apache/spark/commit/fb34cbeeec8a041a6e8a96cc60e962e78116a19f).

That Spark commit explains that the k-way merge comparator only consumes the primary sort key, direction, and null ordering. `sameOrderExpressions` is planner-only metadata, is serialized unnecessarily with every task, and was the vector for the deeply nested `LazyList` `StackOverflowError` addressed by SPARK-58323. The CPU implementation now uses `kWayMergeOrdering` to copy every sort order with `sameOrderExpressions = Seq.empty` before constructing the serialized merge ordering.

The GPU path currently captures the complete `groupPartitions.outputOrdering` in `GpuSorter`. `GpuSortOrderMeta` also treats `sameOrderExpressions` as children, so this can both re-expose the serialization failure and reject GPU replacement because of an unsupported expression that the sorter never evaluates.

Suggested change:

```scala
private val sortedMergeOrdering: Seq[BaseExprMeta[SortOrder]] =
  if (groupPartitions.enableSortedMerge) {
    groupPartitions.outputOrdering
      .map(_.copy(sameOrderExpressions = Seq.empty))
      .map(GpuOverrides.wrapExpr(_, conf, Some(this)))
  } else {
    Seq.empty
  }
```

Please keep `groupInfo.outputOrdering` unchanged for the Spark planner contract; only the ordering passed to the GPU sorter needs this cleanup. Please also add a regression test where the primary sort child is GPU-compatible but `sameOrderExpressions` contains an unsupported expression, and verify that the irrelevant metadata neither causes fallback nor survives in `gpuOutputOrdering`.

File: sql-plugin/src/main/spark420/scala/com/nvidia/spark/rapids/shims/GpuGroupPartitionsExec.scala [hunk 10]
```diff
@@ -133,23 +176,68 @@ case class GpuGroupPartitionsExec(
...
+    val sorter = new GpuSorter(groupInfo.gpuOutputOrdering, output, allMetrics)
+    val targetSize = GpuSortExec.targetSize(conf)
+    val opTime = gpuLongMetric(OP_TIME_LEGACY)
+    val sortTime = gpuLongMetric(SORT_TIME)
+    val outputBatches = gpuLongMetric(NUM_OUTPUT_BATCHES)
+    val outputRows = gpuLongMetric(NUM_OUTPUT_ROWS)
+    coalesced.mapPartitionsWithIndex { case (partitionIndex, batches) =>
+      // Only combining multiple independently sorted streams can invalidate global ordering.
+      if (partitionsNeedingSort(partitionIndex)) {
+        val sorted = GpuOutOfCoreSortIterator(
```
Comment (res-life) [inline sql-plugin/src/main/spark420/scala/com/nvidia/spark/rapids/shims/GpuGroupPartitionsExec.scala:206]: Performance follow-up: reuse the ordering already guaranteed by the child

The CPU implementation leverages the sorted-input contract, but the GPU implementation currently never leverages it. Spark selects the sorted-merge path only when the child reports a non-empty `outputOrdering`. `SortedMergeCoalescedRDD` therefore treats each child partition as an already sorted stream and performs a k-way merge without sorting those inputs again.

Here, the coalesced GPU batches are passed to `GpuOutOfCoreSortIterator`. Its first stage is `GpuSpillableProjectedSortEachBatchIterator`, which calls `appendProjectedAndSort` for every input batch. This re-sorts batches that are already sorted according to the DataSource ordering, discards the main CPU optimization, and is a likely contributor to the 32-stream regression.

This can be addressed in a follow-up without changing the behavior of the widely used existing sort path. Please add an explicit opt-in such as `inputAlreadySorted: Boolean = false`, or a separate `fromSortedInput` factory. Existing callers should retain the current default behavior, while only `GpuGroupPartitionsExec` opts into the sorted-input path. Conceptually:

```scala
private val sortedRuns =
  if (inputAlreadySorted) {
    GpuSpillableProjectedEachBatchIterator(iter, sorter, opTime)
  } else {
    GpuSpillableProjectedSortEachBatchIterator(iter, sorter, opTime, sortTime)
  }
```

The sorted-input iterator must still call `appendProjectedColumns`: computed sort expressions may require temporary projected keys even though the rows are already ordered by those keys. It should skip only `appendProjectedAndSort`, while preserving empty-batch cleanup, `SpillableColumnarBatch` ownership, split-and-retry/OOM handling, spilling, and task-completion cleanup. The call here would then explicitly pass `inputAlreadySorted = true`.

This is safe because `outputOrdering` is a physical-plan contract for each child partition row stream, and coalescing preserves row order within each parent partition. Flattening the streams into batches loses the partition boundary, but every emitted batch remains a sorted run, which is sufficient input for the existing external merge stage.

Please cover at least: one already sorted batch with no local sort work, several interleaved sorted batches producing globally sorted output, a computed sort expression that requires projected columns, and retry/OOM behavior. Ideally the first case should also verify that `sortTime` remains zero.

This optimization removes the redundant per-batch sorts; it does not by itself fix any amplification from the current left-deep pairwise merge or implement a true k-way merge.

After the follow-up, please rerun the exact benchmark used in this PR and report updated medians and relative performance for both 8 streams/key and 32 streams/key, using the same setup and fresh processes. In particular, compare against the currently reported 8-stream GPU 4.940 s vs CPU 7.752 s and 32-stream GPU 35.237 s vs CPU 32.641 s results.

Comment (firestarman) [reply sql-plugin/src/main/spark420/scala/com/nvidia/spark/rapids/shims/GpuGroupPartitionsExec.scala:206]: good idea. 
here is the follow-up issue https://github.com/NVIDIA/cudf-spark/issues/15798

File: sql-plugin/src/main/spark420/scala/com/nvidia/spark/rapids/shims/GpuGroupPartitionsExec.scala (hunk 9 above)
Comment (firestarman) [reply sql-plugin/src/main/spark420/scala/com/nvidia/spark/rapids/shims/GpuGroupPartitionsExec.scala]: nice cathc, updated.

**Comment by firestarman** [comment]
build

**Review by res-life — APPROVED** [review]
LGTM
